scispacy/data_util.py

```diff
@@ -61,18 +61,69 @@ def med_mentions_example_iterator(filename: str) -> Iterator[MedMentionExample]:
         if lines:
             yield process_example(lines)
 
-def read_med_mentions(filename: str):
+def remove_overlapping_entities(sorted_spacy_format_entities: List[Tuple[int, int, str]]
+                                ) -> List[Tuple[int, int, str]]:
     """
-    Reads in the MedMentions dataset into Spacy's
-    NER format.
+    Removes overlapping entities from the entity set, by greedilytaking the longest
+    entity from each overlapping chain. The input list of entities should be sorted
+    and follow the spacy format.
     """
-    examples = []
-    for example in med_mentions_example_iterator(filename):
-        spacy_format_entities = [(x.start, x.end, x.mention_type) for x in example.entities]
-        examples.append((example.text, {"entities": spacy_format_entities}))
-
-    return examples
+    entity_index = 0
+    spacy_format_entities_without_overlap = []
+    stack = []
+    stack_start = 0
+    stack_end = 0
+    while True:
+        if entity_index < len(sorted_spacy_format_entities):
+            current_entity = sorted_spacy_format_entities[entity_index]
+            current_entity_start = current_entity[0]
+            current_entity_end = current_entity[1]
+        else:
+            # if we are after the end of the list, just go through the last iteration to dump the stack
+            current_entity = None
 
+        if current_entity is not None and len(stack) == 0:
+            stack.append(current_entity)
+            stack_start = current_entity_start
+            stack_end = current_entity_end
+        else:
+            if current_entity is not None \
```

Comment:
Hmm this logic is pretty confusing.

can't we do it using a scan over the entities?

```python
sorted_by_start_then_end = sorted(entities, key= lambda x: (x[0], x[1]))
non_overlapping = [sorted_by_start_then_end[0]]

for entity in sorted_by_start_then_end[1:]:
    start, end, _ = entity
	prev_start, prev_end, _ = non_overlapping[-1]

	# case 1: strict non-overlapping.
	if start >= prev_end:
		non_overlapping.append(entity)
	
	# case 2: spans are the same, or new entity is 
	# contained within previous. So skip
	elif start >= prev_start and end <= prev_end:
		continue
	
	# case 3: overlapping, so we need to check which is longer.
	elif start > prev_start and end > prev_end:
		
		# this entity is bigger - pop previous, add this one instead.
		if end - start > prev_end - prev_start:
			non_overlapping.pop()
			non_overlapping.append(entity)
```

Maybe i'm missing some assumption which will make this not work?